### ISSUE TITLE
BFV sec parameter and serialized ciphertext size

### ISSUE THREAD

**Issue by davidhassel** [ibarrond/Pyfhel#142]
Hi,

I have three questions.

Pyfhel version 3.2.1

1. Why is the security level not set to 256 here? The same happens when I try sec=192 and different values on other parameters.

```
from Pyfhel import Pyfhel
HE = Pyfhel()
HE.contextGen(scheme='BFV', n=2**14, t_bits=20, sec=256)
print(HE)
>>> <bfv Pyfhel obj at 0x7f384edcb550, [pk:-, sk:-, rtk:-, rlk:-, contx(n=16384, t=786433, sec=128, qi=[], scale=1.0, )]>
print(HE.sec)
>>> 128
```

2. What determines the serialized ciphertext sizes? Why 32873 in the example below?

```
import numpy as np
from Pyfhel import Pyfhel
HE = Pyfhel()
HE.contextGen(scheme='BFV', n=2**11, t_bits=20, sec=128)
HE.keyGen()

a = np.array([10, 20, 30], dtype=np.int16)
c = HE.encrypt(a)
c_bytes = c.to_bytes()
print(len(c_bytes))
>>> 32873
```

3. What are the downsides of using larger t_bits?

**Comment by AlexanderViand** [comment]
1. That certainly looks like a bug, good catch!
2. The ciphertext is made up of two polynomials, each of which has `n=2**11` coefficients, each of which is `log2(q)` bits wide. Since you're using small parameters, `q < 2**6` (since otherwise you wouldn't get 128-bit security). So there's about 2**18 bits in there if you were to naively serialize. In practice, SEAL does several things, such as using the fact that while the ciphertext is indistinguishable from random in the mathematical structure it exists in, the binary representation has lots of redundancy and so they use compression. They can also sometimes store a seed that generates the second polynomial instead of the full polynomial. So the serialization size might be significantly smaller than the size one would expect.
3. A larger plaintext modulus requires a larger ciphertext modulus (the `q` I mentioned before), so to get the same security, you might have to use larger parameters which slows down operations.

**Comment by davidhassel** [comment]
Thanks for the explanation, much appreciated.

Is there a way to determine or print the ciphertext modulus `q`? Based on the parameters, you say that `q < 2**6`? Why 6? And why not equals?

From my understanding, it is based on the homomorphic encryption standard, but the standard lists much larger `log2(q)`?

**Comment by AlexanderViand** [comment]
Yes, my guess for q was based on the fact that for N=2*11, sec=128bit, Table 2 gives a max log(q) of only 53 bits. That's also why it's a "less than". I think there should be a way to query Pyfhel for the q that was chosen, but let me ping @ibarrond for that :)

**Comment by davidhassel** [comment]
@AlexanderViand, thanks.

I think `HE.total_coeff_modulus_bit_count` in Pyfhel should give `q`. When testing different values and comparing to the standard, I think Pyfhel uses Table 1 with distribution (-1, 1).

But the math does still not add up (at least for me at this time). Using `N=2**11, sec=128` results in `q=54` (not 53). If one would naively serialize, we would get `(2**11)*54*2 = 221184 bits = 27648 bytes`. However, using the serialization `.to_bytes()` with these parameters gives `32873 bytes` (as per the code above).

The following adds up more, but assumes `q=64`.

`(2**11)*64*2 = (2**18) = 262144 bits = 32768 bytes ~= 32873 serialized bytes`

Note that my confusion not only occurs with `N=2**11, sec=128`, but with other parameters (and other `q`) as well. It seems like `N=2**10` and `to_bytes()` gives `16489 bytes`, while `N=2**10` and `to_bytes()` results in `32873` bytes, regardless of the `sec` parameter and chosen `q`. Here it looks like the `q` is always set to 64?

**Comment by AlexanderViand** [comment]
It seems like `to_bytes` does things naively, without the serialization optimizations. Maybe it's because Pyfhel's version of SEAL is compiled without the compression library they use?

The "always 64 bits" thing seems easier to explain: SEAL's internal representation uses standard-sized types (like `uint64_t`) so I guess what you're seeing is just everything rounded up to multiples of 64 bits because that's how SEAL represents it in-memory.

**Comment by davidhassel** [comment]
@AlexanderViand, thanks, makes sense.

It looks like `to_bytes()` in Pyfhel takes an argument `compr_mode` and allows zlib and zstd compression. With zlib and zstd, the ciphertext size was reduced from 32873 to around 31007 and 30921, respectively. Not much, but atleast shows the compression functionality. What is interesting is that the ciphertext now varies a bit between runs.

From the [comments](https://github.com/microsoft/SEAL/blob/main/native/src/seal/serialization.h#L15) in SEAL serialization.h file:

> A type to describe the compression algorithm applied to serialized data. Ciphertext and key data consist of a large number of 64-bit words storing integers modulo prime numbers much smaller than the word size, resulting in a large number of zero bytes in the output. Any compression algorithm should be able to clean up these zero bytes and hence compress both ciphertext and key data.

I think this pretty much sums it up. Thanks again for the help.

**Comment by ibarrond** [comment]
A few quick answers here and there:
1. Well spotted bug! It is now corrected as per commit 5e0a456 (v3.3.0)
2. You can get the total modulus bitcount by running `HE.total_coeff_modulus_bit_count()`, which should correspond to the sum of bitcounts of each moduli qi `sum(HE.qi_sizes)`. Alternatively, in v3.3.0+ you can get the actual modulus chain `HE.qi` and count the bits of each small modulus!
3. As mentioned in #134,  you can now get the size of serialized objects when saving/loading as return value of the respective functions.
4. `save` and `load` functions in Pyfhel are compressed by default (`compr_mode='zstd'`), whereas `to_bytes` and `from_bytes` functions are uncompressed by default (`compr_mode="none"`). In any case, Pyfhel fully supports compression!

Closing this question, feel free to reopen if more related questions pop up

**Comment by davidhassel** [comment]
After looking into this in more detail, there are some more related questions.

Given the same poly modulus degree `n`, a higher security level results in a lower `q`. This means that a higher security level leads to a (theoretical) smaller ciphertext size? How does this add up?

When looking at the speed of various operations, given the same `n`, it looks like a higher security level is faster than using a lower security level. Is this expected? Maybe it has to do with `q` in the previous question?

@ibarrond 
@AlexanderViand
